### DISCUSSION TITLE
How to calculate the storage key for the STATE column family

### DISCUSSION THREAD

**Discussion by helmealex** [paritytech/substrate#11110] [Q&A]
I would like to directly interrogate the node data from the RocksDb database. The data I would like to interrogate is the storage system events from the STATE column family. How is the key calculated for a given hash/block height, the System module and the events storage?

**Comment by bkchr** [comment]
Why do you want to read the raw storage? And not just use the provided methods?

**Reply by helmealex** [reply]
For a boost of performance, I would like to directly index the data, without a middleman. I would like to do something similar to substrate-archive, but in golang, so I need a formula (if possible) to calculate the rocksdb storage key.

**Reply by bkchr** [reply]
CC @cheme @arkpar

**Reply by cheme** [reply]
To directly access data from the state column, you still need to use the https://github.com/paritytech/substrate/blob/297a5e4a548744a07aabb6a8e200f808c54231f5/primitives/trie/src/lib.rs#L185 example of usage can be found in trie_backend_essence.rs from state machine crate.
When accessing directly data with rocksdb api, you will access 'key -> value' that are 'hash(encoded_trie_node)' -> 'value(encoded_trie_node)' (with a prefix in front of the hash for rocksdb, but prefix cannot be transformed to the actual storage key: it is just the begining of it).
So the fastest is certainly to build the triedb over your block root and the underlying database, then iterate from it.

**Reply by ggwpez** [reply]
Did you have a look at [GSRPC](https://github.com/centrifuge/go-substrate-rpc-client) for golang?  
It is not completely up-to-date, but calculating storage keys can be done with it like [here](https://github.com/perun-network/perun-polkadot-backend/blob/30ffc78b7ecd133d6cbb98872f5411a61694bbef/pkg/substrate/api.go#L64) or [here](https://github.com/perun-network/perun-polkadot-backend/blob/30ffc78b7ecd133d6cbb98872f5411a61694bbef/pkg/substrate/timeout.go#L103).

**Reply by helmealex** [reply]
So, from what I understand, you get the state root of the block, get the data from the sate column for that hash and the data is the encoded trie node for the root trie node. After that let's suppose I want to access the storage events. How? I attach the encoded keys for storage events to what key? Or i have to iterate through the child of the root trie node?

**Reply by helmealex** [reply]
> Did you have a look at [GSRPC](https://github.com/centrifuge/go-substrate-rpc-client) for golang? It is not completely up-to-date, but calculating storage keys can be done with it like [here](https://github.com/perun-network/perun-polkadot-backend/blob/30ffc78b7ecd133d6cbb98872f5411a61694bbef/pkg/substrate/api.go#L64) or [here](https://github.com/perun-network/perun-polkadot-backend/blob/30ffc78b7ecd133d6cbb98872f5411a61694bbef/pkg/substrate/timeout.go#L103).

I know, but these are the keys used to interrogate the RPC endpoint. You cannot use them to directly access the underlying rocksdb database.

**Reply by cheme** [reply]
If you want a storage event you need the key to the event value and access the stored event sfor this key  (get) from the instantiated triedb struct (events being stored are those of the previous block).
In itself events are stored at https://github.com/paritytech/substrate/blob/297a5e4a548744a07aabb6a8e200f808c54231f5/frame/system/src/lib.rs#L597 , and the trie key is calculated by the pallet storage macro macro (might be something like twox(name_of_pallet) ++ twox(name_of _field).
I know @shawntabrizi got some great javascrit to get the trie keys of the runtime storage, also found these post he writes which are quite nice https://www.shawntabrizi.com/substrate/querying-substrate-storage-via-rpc/ and  https://www.shawntabrizi.com/substrate/transparent-keys-in-substrate/ (rpc for state are a good way to access storage, and basically if you want to go faster you can only just skip the rpc part and call directly the code in rust (calling the code behind the rpc directly may be rather good (the bigger cost is still accessing trie node from the db)).
